### PR TITLE
feat(content-blog): parse date from middle of file path

### LINKED ISSUE #5554: Blog date is no longer parsed correctly when blog has subfolders
### Prerequisites

- [x] I'm using the latest version of Docusaurus.
- [x] I have tried the `npm run clear` or `yarn clear` command.
- [x] I have tried `rm -rf node_modules yarn.lock package-lock.json` and re-installing packages.
- [ ] I have tried creating a repro with https://new.docusaurus.io
- [x] I have read the console error message carefully (if applicable)

### Description

Our blog file structure contains subfolders for better organization. We have hundreds of blogposts from different internal teams so this is vital for us:
```
/blog
  /team-a
    YYYY-MM-DD-name.md
    YYYY-MM-DD-name.md
  /team-b
    YYYY-MM-DD-name.md
    YYYY-MM-DD-name.md
```

Starting from 2.0.0-beta.5 docusaurus-plugin-content-blog is no longer able to parse blog date for such folder organization. It changed in the following PR [#5309](https://github.com/facebook/docusaurus/pull/5309/files). I understand that our blog folder structure use case is not officially supported but please consider adding support for it. 

My understanding is that previously date was parsed from filename `path.basename(blogSourceFile)`, but now it is parsed from entire filepath to support blog folders like `2021-05-12-announcing-docusaurus-two-beta/index.md`.

I think it is relatively low effort to support parsing dates for all currently supported conventions + allow subfolders for better organization and I would be happy to contribute PR for that:
1. `2021-05-28-my-blog-post-title.md`
2. `2021-05-28-my-blog-post-title/index.md`
3. `team-a/2021-05-28-my-blog-post-title.md`
4. `team-a/2021-05-28-my-blog-post-title/index.md`

### Have you read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues)?

Yes

## Steps to reproduce

Already explained above.

## Your environment

N/A

## Reproducible demo

N/A

### PR DESCRIPTION
**Pull request by Josh-Cena**
## Motivation

Fix #5554.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Unit test + dogfood

### CODE REVIEW

File: packages/docusaurus-plugin-content-blog/src/blogUtils.ts [hunk 1]
```diff
@@ -74,12 +74,11 @@ export function parseBlogFileName(
...
   const dateFilenameMatch = blogSourceRelative.match(DATE_FILENAME_REGEX);
   if (dateFilenameMatch) {
-    const dateString = dateFilenameMatch.groups!.date!;
-    const text = dateFilenameMatch.groups!.text!;
+    const {folder, text, date: dateString} = dateFilenameMatch.groups!;
     // Always treat dates as UTC by adding the `Z`
     const date = new Date(`${dateString}Z`);
     const slugDate = dateString.replace(/-/g, '/');
-    const slug = `/${slugDate}/${text}`;
+    const slug = `/${slugDate}/${folder}${text}`;
```
Comment (slorber) [inline packages/docusaurus-plugin-content-blog/src/blogUtils.ts:81]: so folder always ends with a /? I think this a bit confusing but why not, as long as the test covers it...

**Review by slorber — APPROVED** [review]
LGTM thanks 👍 

Maybe we can wait a few days for @kanoshin feedback?

File: packages/docusaurus-plugin-content-blog/src/blogUtils.ts (hunk 1 above)
Comment (Josh-Cena) [reply packages/docusaurus-plugin-content-blog/src/blogUtils.ts:81]: Well, not always, as is the case for `team-a-2021-05-12-hey/announcing-docusaurus-two-beta.md` => `/2021/05/12/team-a-hey/announcing-docusaurus-two-beta`... But in terms of spec I can't think of any better heuristic... Maybe the date string should always be preceded by `/`?

**Comment by Josh-Cena** [comment]
Yes, we'll wait for more feedback

File: packages/docusaurus-plugin-content-blog/src/blogUtils.ts (hunk 1 above)
Comment (slorber) [reply packages/docusaurus-plugin-content-blog/src/blogUtils.ts:81]: 🤷‍♂️ that seems good enough for now, we'll see if current regexp produce any weird issue

**Comment by Josh-Cena** [comment]
Hi @kanoshin! We are still waiting for your feedback on this feature, since you presented the use-case first. We will probably merge it this week.

**Comment by Josh-Cena** [comment]
Going to merge this since we don't get any feedback... I think this feature is good to go.

**Pull request merged**